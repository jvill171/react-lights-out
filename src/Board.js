import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game.
 *   Note: chanceLightStartsOn must be between 0 (inclusive) and 1 (exclusive)
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows = 5, ncols = 5, chanceLightStartsOn = .5 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];

    // Create array-of-arrays of true/false values
    // Rows
    for(let rIdx=0; rIdx<nrows; rIdx++){
      const row = []
      // Cells in rows
      for(let cIdx=0; cIdx<ncols; cIdx++){
        // If randomly generated number < chanceLightStartsOn, cell starts on 
        row.push( Math.random() < chanceLightStartsOn ? true : false )
      }
      initialBoard.push(row)
    }
    return initialBoard;
  }

  // Check the board in state to determine whether the player has won.
  function hasWon() {
    return board.every(r => r.every(c=> c===true ))
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it
        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // Make a (deep) copy of the oldBoard
      const deepCopyBoard =  oldBoard.map(rows => [...rows])

      // In the copy, flip this cell .....
      flipCell(y, x, deepCopyBoard)
      // ..... and the cells around it
      flipCell(y, x+1, deepCopyBoard)
      flipCell(y, x-1, deepCopyBoard)
      flipCell(y+1, x, deepCopyBoard)
      flipCell(y-1, x, deepCopyBoard)

      // TODO: return the copy
      return deepCopyBoard;
    });
  }

  return (
    // if the game is won, just show a winning msg & render nothing else
    hasWon() ?  <h1>WINNER</h1> :
      // make table board
      <table>
        {board.map((row, rIdx) => {
          return (
            <tr>
            { row.map((cell, cIdx) => {
              return <Cell flipCellsAroundMe={()=>flipCellsAround(`${rIdx}-${cIdx}`)} isLit={board[rIdx][cIdx]}/>
            })}
            </tr>
            )
        })}
      </table>
    )
  
}

export default Board;
