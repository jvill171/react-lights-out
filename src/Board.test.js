import { render, fireEvent, queryAllByText } from "@testing-library/react"
import Board from "./Board"

// TEST_BOARD will always be a 100% off board, with all cells turned unlit.
// 3 x 3 board makes checking for a win easy: Click center & corners 1 time
// 0% chance of lights on makes snapshots reliable
const TEST_BOARD = {
    nrows: 3,
    ncols: 3,
    chanceLightStartsOn: 0
}

// Smoke test
it("renders without crashing", ()=>{
    // Uses default values from Board component
    render(<Board/>)
})

// Snapshot test
it("matches snapshot", ()=>{
    let { nrows, ncols, chanceLightStartsOn } = TEST_BOARD
    const {asFragment} = render(
        <Board
            nrows={nrows}
            ncols={ncols}
            chanceLightStartsOn={chanceLightStartsOn}
        />
    )
    expect(asFragment()).toMatchSnapshot()
})

// Cell Clicking
it("flips itself and its 4 neighboring cells", ()=>{
    let { nrows, ncols, chanceLightStartsOn } = TEST_BOARD
    const { container } = render(
        <Board
            nrows={nrows}
            ncols={ncols}
            chanceLightStartsOn={chanceLightStartsOn}
        />
    );

    const centerCell = container.querySelectorAll('.Cell')[4]
    fireEvent.click(centerCell);
    // Check only 5 cells lit up
    expect(container.querySelectorAll('.Cell-lit').length).toEqual(5)

    // List of cells expected to NOT be lit
    const unLit = [0, 2, 6, 8]
    const clickedBoard = container.querySelectorAll('.Cell')
    // Ensure only the correct cells lit up
    for(let i = 0; i<(nrows * ncols); i++){
        // Current status of the cell after the click
        const isCellLit = clickedBoard[i].classList.contains('Cell-lit')
        // What should the status of that cell be?
        const shouldBeLit = !unLit.includes(i)
        expect(isCellLit).toEqual(shouldBeLit)
    }
})

it("removes the game board and displays a winning message when the game is won", ()=>{
    let { nrows, ncols, chanceLightStartsOn } = TEST_BOARD
    const { container, debug } = render(
        <Board
            nrows={nrows}
            ncols={ncols}
            chanceLightStartsOn={chanceLightStartsOn}
        /> 
    );
  
    expect(container.querySelector('table')).toBeInTheDocument();
    expect(queryAllByText(container, "WINNER")).toEqual([]); 

    const boardCells = container.querySelectorAll('.Cell')
    // Click the the corner cells and the center cell,
    // winning in a 3x3 board of fully unlit cells
    fireEvent.click(boardCells[0]); // top-left corner
    fireEvent.click(boardCells[2]); // top-right corner
    fireEvent.click(boardCells[4]); // center
    fireEvent.click(boardCells[6]); // bottom left corner
    fireEvent.click(boardCells[8]); // bottom-right corner

    expect(container.querySelector('table')).not.toBeInTheDocument();
    expect(queryAllByText(container, "WINNER")[0]).toBeInTheDocument();
    
})