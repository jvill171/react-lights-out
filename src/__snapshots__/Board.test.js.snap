// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="Cell"
        />
        <td
          class="Cell"
        />
        <td
          class="Cell"
        />
      </tr>
      <tr>
        <td
          class="Cell"
        />
        <td
          class="Cell"
        />
        <td
          class="Cell"
        />
      </tr>
      <tr>
        <td
          class="Cell"
        />
        <td
          class="Cell"
        />
        <td
          class="Cell"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
